atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/ClientLogEvents.java

```diff
@@ -94,10 +101,53 @@ default CommitUpdate toCommitUpdate(
         verifyReturnedEventsEnclosesTransactionVersions(startVersion.version() + 1, endVersion.version());
 
         LockEventVisitor eventVisitor = new LockEventVisitor(commitInfo.map(CommitInfo::commitLockToken));
-        Set<LockDescriptor> locksTakenOut = events().events().stream()
+
+        Set<LockDescriptor> locksFromLockWatchCreatedEvents = events().events().stream()
+                .flatMap(event -> event.accept(LockWatchCreatedEventVisitor.INSTANCE).stream())
+                .collect(Collectors.toUnmodifiableSet());
+        List<LockEvent> lockEvents = events().events().stream()
                 .flatMap(event -> event.accept(eventVisitor).stream())
-                .collect(ImmutableSet.toImmutableSet());
-        return CommitUpdate.invalidateSome(locksTakenOut);
+                .collect(Collectors.toUnmodifiableList());
+        Set<LockDescriptor> locksFromLockEvents = lockEvents.stream()
+                .map(LockEvent::lockDescriptors)
+                .flatMap(Set::stream)
+                .collect(Collectors.toUnmodifiableSet());
+
+        /*
+         * Metadata aggregation semantics for a lock descriptor L:
+         * - If L appears in a LockWatchCreatedEvent, L will be excluded from the aggregation result.
+         * - Otherwise, if there is a lock event that contains L, but L does not have metadata associated in this lock
+         *   event, L is also excluded from the aggregation result.
+         * - Otherwise, the metadata objects associated with L are listed in the order of the lock events they appear
+         *   in.
+         */
```

Comment:
I *think* these are the semantics we want/need, but this is something that has potential to change once we start using metadata.